source/Core/Drivers/FUSB302/policy_engine.cpp

```diff
@@ -617,6 +619,18 @@ void PolicyEngine::PPSTimerCallback() {
   }
 }
 
+bool PolicyEngine::NegotiationTimeoutReached() {
+    if (systemSettings.PDNegTimeout == 0){
+	return false;
+    }
+
+    if (xTaskGetTickCount() > (TICKS_SECOND/10 * systemSettings.PDNegTimeout)) {
```

Comment:
You need to make the timeout for the waiting for capabilities state dynamic, don't just patch over it or else pd negotiation may still continue in the background.

Reply:
> You need to make the timeout for the waiting for capabilities state dynamic, don't just patch over it or else pd negotiation may still continue in the background.

I did it because I have thought if it continues during the background and is successful, then the result of PD negotiation would be much better than QC itself.

Is it a bad idea?

Maybe if QC3 pulses has started then it drops ongoing PD initialization, I don't know.
That case, so if PD negotiation is slow and QC3 pulses break remaining part of negotiation, then it should definitively go to the state machine, but is that the case?

Reply:
That is indeed the case on a charger and a dock I have to hand, where as soon as you start to mess with the data pins they lock the port to 5v.

Also it's just for clarity sake when debugging a user having an issue, far easier of it's known "once QC starts, PD stops"

Reply:
> This is a standalone class, you can't rely on general tick counter, please reference this based on timing out during the wait for capabilities state in the state machine

I am afraid I don't know how to do it in a proper way, I thought the way PPSTimerCallback() uses may be good.

One thing is it should put state machine to appropriate state, but another one is I would like to wait for given timeout, regardless of what state the PD negotiation is.
Also, I am not familiar with RTOS yet, so I don't know if I set the state of machine elsewhere, then it may also change the state machine in one of the states (still pending) or... how should I do it.